File: scala/triangle/src/main/scala/Triangle.scala
case class Triangle(x: Int, y: Int, z: Int) {
  def equalSides = Set(x, y, z).size

  def isValid = List(x, y, z).sorted match {
    case x :: y :: z :: Nil => x + y > z
  }

  def triangleType =
    if (isValid)
      equalSides match {
        case 2 => TriangleType.Isosceles
        case 1 => TriangleType.Equilateral
        case _ => TriangleType.Scalene
      }
    else TriangleType.Illogical
}

object TriangleType extends Enumeration {
  type TriangleType = Value
  val Equilateral, Isosceles, Scalene, Illogical = Value
}

File: fsharp/triangle/Triangle.fs
module Triangle

let private isValid triangle =
    let [x; y; z] = triangle    
    x > 0. && y > 0. && z > 0. &&
    x + y >= z && x + z >= y && y + z >= x

let private distinctSides triangle = triangle |> List.distinct |> List.length

let equilateral triangle = isValid triangle && distinctSides triangle = 1

let isosceles triangle = isValid triangle && distinctSides triangle <= 2

let scalene triangle = isValid triangle && distinctSides triangle = 3